src/lib/bm_ncp/ncp_uart.cpp

```diff
@@ -446,6 +559,41 @@ static void ncpRXProcessor(void *parameters) {
   }
 }
 
+static void ncpBaudRateDiscovery(void) {
+  uint8_t baud_idx = 0;
+  static const uint8_t null_buf = 0;
+
+  ncp_baud_rate_discovery_lock = xSemaphoreCreateBinary();
+  configASSERT(ncp_baud_rate_discovery_lock);
+
+  while (!ncp_ctx.baud_found && baud_idx < array_size(acceptable_bauds)) {
+    // Reset buffers just incase anything was sent at a different baud rate
+    ncpRXCurrBuff = 0;
+    ncpRXBuffIdx = 0;
+    serialSetBaudRate(ncpSerialHandle, acceptable_bauds[baud_idx]);
+    bm_serial_tx(BM_SERIAL_ACK, &null_buf, sizeof(uint8_t));
+    ncp_negotiate_baud_rate(921600);
+    baud_idx++;
+    if (xSemaphoreTake(ncp_baud_rate_discovery_lock, portMAX_DELAY) != pdTRUE) {
+      printf("Could not take baud rate discovery semaphore\n");
+      break;
+    }
+  }
+
+  if (!ncp_ctx.baud_found) {
+    serialSetBaudRate(ncpSerialHandle, 115200);
+    // Send 0 to reset receiving buffer on spotter and reset our buffers just
+    // incase anything was sent at a different baud rate
+    ncpRXCurrBuff = 0;
+    ncpRXBuffIdx = 0;
+    bm_serial_tx(BM_SERIAL_ACK, &null_buf, sizeof(uint8_t));
+    printf("Could not find a baudrate to communicate to spotter with\n");
+  }
+
+  vSemaphoreDelete(ncp_baud_rate_discovery_lock);
+  ncp_baud_rate_discovery_lock = NULL;
```

Comment:
```suggestion
  // Prevent race condition with semaphore usage in other tasks
  SemaphoreHandle_t tmp_sem = ncp_baud_rate_discovery_lock;
  ncp_baud_rate_discovery_lock = NULL;
  vSemaphoreDelete(tmp_sem);
```

My suggestion fixes a race condition here with the usage of `ncp_baud_rate_discovery_lock` in interrupts or other tasks. If we delete the semaphore, and then the scheduler stops this task before the static variable is set to NULL, we could call `xSemaphoreGive(ncp_baud_rate_discovery_lock)` on the deleted semphore (whose memory has been freed) in either `baud_rate_reply_cb` or `baud_rate_negotiate_timer_handle`.

Reply:
Ahhh very good call out! I will go ahead and make that update, thanks!